pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -132,6 +147,163 @@ func storageMatches(set *apps.StatefulSet, pod *v1.Pod) bool {
+	policy := apps.RetainPersistentVolumeClaimDeletePolicyType
+	if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) && set.Spec.PersistentVolumeClaimDeletePolicy != nil {
+		policy = *set.Spec.PersistentVolumeClaimDeletePolicy
+	}
+	return policy
+}
+
+// claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
+// PVC deletion policy for the StatefulSet.
+func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	policy := getPersistentVolumeClaimDeletePolicy(set)
+	switch policy {
+	default:
+		klog.Errorf("Unknown policy %v; treating as Retain", set.Spec.PersistentVolumeClaimDeletePolicy)
+		fallthrough
+	case apps.RetainPersistentVolumeClaimDeletePolicyType:
+		if hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnStatefulSetDeletionOnlyPersistentVolumeClaimDeletePolicyType:
+		if !hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownOnlyPersistentVolumeClaimDeletePolicyType:
+		if hasOwnerRef(claim, set) {
+			return false
+		}
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownAndStatefulSetDeletionPersistentVolumeClaimDeletePolicyType:
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		// If a pod is scaled down, there should be no set ref and a pod ref;
+		// if the pod is not scaled down it's the other way around.
+		if podScaledDown == hasOwnerRef(claim, set) {
+			return false
+		}
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	}
+	return true
+}
+
+// updateClaimOwnerRefForSetAndPod updates the ownerRefs for the claim according to the deletion policy of
+// the StatefulSet. Returns true if the claim was changed and should be updated and false otherwise.
+func updateClaimOwnerRefForSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	needsUpdate := false
+	// Sometimes the version and kind are not set pod.TypeMeta. These are necessary for the ownerRef.
+	// TODO: there must be a better way to do this other than hardcoding the pod version?
+	podMeta := pod.TypeMeta
+	if podMeta.APIVersion == "" {
+		podMeta.APIVersion = "v1"
+	}
+	if podMeta.Kind == "" {
+		podMeta.Kind = "Pod"
+	}
+	policy := getPersistentVolumeClaimDeletePolicy(set)
+	switch policy {
+	default:
+		klog.Errorf("Unknown policy %v, treating as Retain", set.Spec.PersistentVolumeClaimDeletePolicy)
+		fallthrough
+	case apps.RetainPersistentVolumeClaimDeletePolicyType:
+		needsUpdate = removeOwnerRef(claim, set) || needsUpdate
+		needsUpdate = removeOwnerRef(claim, pod) || needsUpdate
+	case apps.DeleteOnStatefulSetDeletionOnlyPersistentVolumeClaimDeletePolicyType:
+		needsUpdate = setOwnerRef(claim, set, &set.TypeMeta) || needsUpdate
```

Comment:
Shouldn't the Sts ref be a controllerRef?

Reply:
I thought an ownerRef was sufficient. TBH there doesn't seem to be a good definition of when controllerRefs should be used, though. For example, why hasn't there ever been a controllerRef on statefulset PVCs?

I'm happy to admit that I got this wrong.

Reply:
sts (as written today) doesn't own its PVCs. GC and cascading deletion intentionally are not applied to PVCs. The controllerRefs basically default the behavior of blocking owner deletion based on children. The question is should the sts be collected after all PVCs have been deleted or should the PVCs just be garbaged collected on best effort. Have we thought through what happens when a user deletes a sts and recreates a sts with the same name in this case?